dev/cuda/adamw.cu

```diff
@@ -121,8 +271,29 @@ void adamw_dispatch2(float* params_memory, const float* grads_memory, float* m_m
     cudaCheck(cudaGetLastError());
 }
 
+// version 3: Using floatX to be able to reduce amount data type size for params and grads
+void adamw_dispatch3(floatX* params_memory, const floatX* grads_memory, float* m_memory, float* v_memory, long num_parameters,
+                     float learning_rate, float beta1, float beta2, float beta1_correction, float beta2_correction, float eps, float weight_decay) {
+    unsigned int block_size = 512;
+    unsigned int num_blocks = ceil_div(num_parameters, (long) block_size);
+    adamw_kernel3<<<num_blocks, block_size>>>(params_memory, grads_memory, m_memory, v_memory, num_parameters,
+                                              learning_rate, beta1, beta2, beta1_correction, beta2_correction, eps, weight_decay, 1337);
+    cudaCheck(cudaGetLastError());
+}
+
+// version 4: using data packing to reduce the amount of amount of memory calls
+void adamw_dispatch4(floatX* params_memory, const floatX* grads_memory, float* m_memory, float* v_memory, long num_parameters,
+                     float learning_rate, float beta1, float beta2, float beta1_correction, float beta2_correction, float eps, float weight_decay) {
+    unsigned int block_size = 512;
+    assert(num_parameters % 4 == 0 && f128::size <= x128::size); // asserting here to not require bounds check in kernel
```

Comment:
f128::size <= x128::size is a compile-time property, best make that a static_assert inside the actual kernel